File: modules/expo-mlkit-face-detection/ios/ExpoMLKitFaceDetectorOptions.swift
//
//  ExpoMLKitFaceDetectorOptions.swift
//  ExpoMLKitFaceDetection
//
//  Created by Trevor Coleman on 2023-07-28.
//

import ExpoModulesCore
import Foundation
import MLKitFaceDetection

public struct ExpoMlKitFaceDetectorOptionsRecord: Record {

    public init() {}

    @Field
    var performanceMode: String = "fast"
    @Field
    var landmarkMode: Bool? = nil
    @Field
    var contourMode: Bool? = nil
    @Field
    var classificationMode: Bool? = nil
    @Field
    var minFaceSize: Float? = nil
    @Field
    var isTrackingEnabled: Bool? = nil

}

public class ExpoMLKitFaceDetectorOptions {

    var performanceMode: FaceDetectorPerformanceMode = .fast
    var landmarkMode: FaceDetectorLandmarkMode = .none
    var contourMode:FaceDetectorContourMode = .none
    var classificationMode:FaceDetectorClassificationMode = .none
    var minFaceSize: CGFloat = 0.1
    var isTrackingEnabled:Bool = false


    init(performanceMode: FaceDetectorPerformanceMode?, landmarkMode: FaceDetectorLandmarkMode?, contourMode: FaceDetectorContourMode?, classificationMode: FaceDetectorClassificationMode?, minFaceSize: CGFloat?, isTrackingEnabled: Bool?) {
        self.performanceMode = performanceMode  ?? .fast
        self.landmarkMode = landmarkMode ?? .none
        self.contourMode = contourMode  ?? .none
        self.classificationMode = classificationMode  ?? .none
        self.minFaceSize = minFaceSize ?? 0.1
        self.isTrackingEnabled = isTrackingEnabled ?? false
    }

    convenience init(record: ExpoMlKitFaceDetectorOptionsRecord) {
        self.init(
            performanceMode: record.performanceMode == "accurate"
            ? FaceDetectorPerformanceMode.accurate
            : FaceDetectorPerformanceMode.fast,
            landmarkMode: record.landmarkMode == true
            ? FaceDetectorLandmarkMode.all
            : FaceDetectorLandmarkMode.none,
            contourMode: record.contourMode == true
            ? FaceDetectorContourMode.all
            : FaceDetectorContourMode.none,
            classificationMode: record.classificationMode == true
            ? FaceDetectorClassificationMode.all
            : FaceDetectorClassificationMode.none,
            minFaceSize: CGFloat(record.minFaceSize ?? 0.1),
            isTrackingEnabled: record.isTrackingEnabled)
    }

    var record: ExpoMlKitFaceDetectorOptionsRecord {
        let record = ExpoMlKitFaceDetectorOptionsRecord()
        record.performanceMode = self.performanceMode == .accurate ? "accurate" : "fast"
        record.landmarkMode = self.landmarkMode == .all
        record.contourMode = self.contourMode == .all
        record.classificationMode = self.classificationMode == .all
        record.minFaceSize = Float(self.minFaceSize)
        record.isTrackingEnabled = self.isTrackingEnabled
        return record
    }

    var options: FaceDetectorOptions {
        let options = FaceDetectorOptions()
        options.performanceMode = self.performanceMode
        options.landmarkMode = self.landmarkMode
        options.contourMode = self.contourMode
        options.classificationMode = self.classificationMode
        options.minFaceSize = self.minFaceSize
        options.isTrackingEnabled = self.isTrackingEnabled
        return options
    }
}

File: modules/expo-mlkit-face-detection/android/src/main/java/red/infinite/expomlkit/facedetection/ExpoMLKitFaceDetectorOptions.kt
//
//  ExpoMLKitFaceDetectorOptions.kt
//  ExpoMLKitFaceDetection
//
//  Created by Trevor Coleman on 2023-07-28.
//

import com.google.mlkit.vision.face.FaceDetectorOptions
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

data class ExpoMLKitFaceDetectorOptionsRecord(
    @Field var performanceMode: String = "fast",
    @Field var landmarkMode: Boolean? = null,
    @Field var contourMode: Boolean? = null,
    @Field var classificationMode: Boolean? = null,
    @Field var minFaceSize: Float? = null,
    @Field var isTrackingEnabled: Boolean? = null
) : Record

class ExpoMLKitFaceDetectorOptions(
    performanceMode: String? = null,
    landmarkMode: Boolean? = null,
    contourMode: Boolean? = null,
    classificationMode: Boolean? = null,
    minFaceSize: Float? = null,
    isTrackingEnabled: Boolean? = null
) {

    var performanceMode = performanceMode ?: "fast"
    var landmarkMode = landmarkMode ?: false
    var contourMode = contourMode ?: false
    var classificationMode = classificationMode ?: false
    var minFaceSize = minFaceSize ?: 0.1f
    var isTrackingEnabled = isTrackingEnabled ?: false

    constructor(record: ExpoMLKitFaceDetectorOptionsRecord) : this(
        performanceMode = record.performanceMode,
        landmarkMode = record.landmarkMode,
        contourMode = record.contourMode,
        classificationMode = record.classificationMode,
        minFaceSize = record.minFaceSize,
        isTrackingEnabled = record.isTrackingEnabled
    )

    val record: ExpoMLKitFaceDetectorOptionsRecord
        get() {
            return ExpoMLKitFaceDetectorOptionsRecord(
                performanceMode = this.performanceMode,
                landmarkMode = this.landmarkMode,
                contourMode = this.contourMode,
                classificationMode = this.classificationMode,
                minFaceSize = this.minFaceSize,
                isTrackingEnabled = this.isTrackingEnabled
            )
        }

    val options: FaceDetectorOptions
        get() {
            val optionsBuilder = FaceDetectorOptions.Builder()
                .setPerformanceMode(
                    if (this.performanceMode == "accurate") FaceDetectorOptions.PERFORMANCE_MODE_ACCURATE else FaceDetectorOptions.PERFORMANCE_MODE_FAST
                )
                .setLandmarkMode(
                    if (this.landmarkMode) FaceDetectorOptions.LANDMARK_MODE_ALL else FaceDetectorOptions.LANDMARK_MODE_NONE
                )
                .setContourMode(
                    if (this.contourMode) FaceDetectorOptions.CONTOUR_MODE_ALL else FaceDetectorOptions.CONTOUR_MODE_NONE
                )
                .setClassificationMode(
                    if (this.classificationMode) FaceDetectorOptions.CLASSIFICATION_MODE_ALL else FaceDetectorOptions.CLASSIFICATION_MODE_NONE
                )
                .setMinFaceSize(this.minFaceSize)

            if (this.isTrackingEnabled) {
                optionsBuilder.enableTracking()
            }

            return optionsBuilder.build()
        }
}
